fake_tensor.py

```diff
@@ -0,0 +1,162 @@
+import torch
+
+from base_tensor import BaseTensor
+from torch.testing._internal.common_utils import (
+    run_tests,
+    TestCase,
+)
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.fx.operator_schemas import normalize_function
+from typing import Union
+import unittest
+
+aten = torch.ops.aten
+
+
+# Meta tensors give you the ability to run PyTorch code without having to
+# actually do computation through tensors allocated on a `meta` device.
+# Because the device is `meta`, meta tensors do not model device propagation.
+# FakeTensor extends MetaTensors to also carry an additional `fake_device`
+# which tracks devices that would have been used.
+
+
+class FakeTensor(BaseTensor):
+    fake_device: torch.device
+
+    @staticmethod
+    def __new__(cls, elem, device):
+        return super().__new__(cls, elem)
+
+    def __init__(self, elem, device: Union[torch.device, str]):
+        # elem does not need to be recorded, because FakeTensor *is a* elem
+        meta_device = torch.device(type="meta")
+        assert elem.device == meta_device
+        assert device != meta_device
+        self.fake_device = (
+            device if isinstance(device, torch.device) else torch.device(device)
+        )
+
+    @staticmethod
+    def from_tensor(t):
+        existing_device = t.device
+        return FakeTensor(t.to(device="meta"), existing_device)
+
+    @property
+    def device(self):
+        return self.fake_device
```

Comment:
As mentioned in chats, this would be insufficient for reporting CPU/CUDA device in C++, which matters if you're sending this tensor through C++ code that does device tests. This might get us pretty far for now though, so I wouldn't block on it.